### PR TITLE
Various fixes

### PR DESCRIPTION
**Pull request by yannikschaelte**
In particular: 
* Introduce constants file currently holding path references
* Use those path constants consistently throughout the scripts
* Define a working environment base folder (ENV SOLVERSTUDY_DIR_BASE)
* Tidy up and speed up JWS model and simulation download (download and store only once)
* Fix problem with repeated download of JWS models
* Fix problem of wrong JWS sbml model slugs

### CODE REVIEW

File: .gitignore [hunk 1]
```diff
@@ -1,4 +1,7 @@
+__pycache__
+.ipynb_checkpoints
+.idea
+SolverStudyWork
+SolverStudyCache
+SolverStudyTest
 *.DS_Store
-Python_Scripts/.idea/*
-Python_Scripts/__pycache__/*
```
Comment (paulstapor) [inline .gitignore:3]: Mein Gefühl sagt mir, dass wir diesen "Python_Scripts"-Ordner vorneweg schon noch brauchen

File: Python_Scripts/C.py [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+
+###############################################################################
+# Base folder
+
+# The simulation scripts write their output to `DIR_WORK_BASE` by default.
+DIR_WORK_BASE = '../SolverStudyWork'
+# The output of the study has been conserved in `DIR_CACHE_BASE`.
+DIR_CACHE_BASE = '../SolverStudyCache'
+# The `DIR_TEST_BASE` folder contains a smaller test model set
+DIR_TEST_BASE = '../SolverStudyTest'
```
Comment (paulstapor) [inline Python_Scripts/C.py]: How about using `os.path.abspath` here? Little more effort, but fail-safe with respect to "where do you run your script from"...

File: Python_Scripts/C.py [hunk 3]
```diff
@@ -0,0 +1,57 @@
+"""Constant definitions."""
+
+import os
+
+###############################################################################
+# Base folder
+
+# The simulation scripts write their output to `DIR_WORK_BASE` by default.
+DIR_WORK_BASE = '../SolverStudyWork'
```
Comment (paulstapor) [inline Python_Scripts/C.py]:
```suggestion
DIR_WORK_BASE = os.path.abspath('../SolverStudyWork')
```

File: Python_Scripts/C.py [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+# The output of the study has been conserved in `DIR_CACHE_BASE`.
+DIR_CACHE_BASE = '../SolverStudyCache'
```
Comment (paulstapor) [inline Python_Scripts/C.py]:
```suggestion
DIR_CACHE_BASE = os.path.abspath('../SolverStudyCache')
```

File: Python_Scripts/C.py (hunk 2 above)
Comment (paulstapor) [inline Python_Scripts/C.py]:
```suggestion
DIR_TEST_BASE = os.path.abspath('../SolverStudyTest')
```

File: Python_Scripts/compareStateTrajectories_JWS_2.py [hunk 5]
```diff
@@ -136,17 +148,22 @@
...
+                                    data=np.zeros((1, 1)))
 
                                 # single error
                                 for iCol in column_names:
                                     for iRow in range(0, amount_row):
-                                        rel_tol = abs((df_state_trajectory.at[iRow, iCol] - tsv_file.at[iRow,iCol]) /
-                                                      df_state_trajectory.at[iRow, iCol])
+                                        rel_tol = abs(
+                                            (df_state_trajectory.at[iRow, iCol] - tsv_file.at[iRow,iCol]) /
+                                             df_state_trajectory.at[iRow, iCol])
```
Comment (paulstapor) [inline Python_Scripts/compareStateTrajectories_JWS_2.py:166]: Aua... Aua... Aua!
Zeilen- und Spaltenweise durchiterieren, um einen relativen Error zu berechnen, wo man einfach den kompletten Numpy-array hätte hernehmen können... Das geht auf die Performance, huiuiui... :turtle:

File: Python_Scripts/compareStateTrajectories_JWS_2.py [hunk 6]
```diff
@@ -136,17 +148,22 @@
...
                                         abs_tol = abs(df_state_trajectory.at[iRow, iCol] - tsv_file.at[iRow,iCol])
                                         if rel_tol <= rel_error or abs_tol <= abs_error:
                                             df_single_error.at[iRow, iCol] = 1
```
Comment (paulstapor) [inline Python_Scripts/compareStateTrajectories_JWS_2.py:169]: Auch die Logik dieses Trajektorienvergleichs ist... Sagen wir für eine Art Debugging-Modus okay. Aber für das finale Repository eher overkill...

File: Python_Scripts/compareStateTrajectories_JWS_1.py [hunk 7]
```diff
@@ -3,36 +3,210 @@
...
+        DIR_MODELS_SEDML, iModel, iModel + '.sedml')
+    sbml_path = os.path.join(
+        DIR_MODELS_SEDML, iModel, 'sbml_models',
+        iFile + '.sbml')
+    BioModels_path = DIR_MODELS_BIOMODELS
+
+    if os.path.exists(os.path.join(BioModels_path, iModel)):
+        print("Model is not part of JWS-database")
+    else:
+        # check if 'all_settings' works
```
Comment (paulstapor) [inline Python_Scripts/compareStateTrajectories_JWS_1.py:111]: Here, you've cut out some logic... Any statement on why this doesn't change the code? Do You circumvent the whole JSON thingi here?

File: Python_Scripts/compareStateTrajectories_JWS_1.py [hunk 8]
```diff
@@ -3,36 +3,210 @@
...
+        # open sbml file
+        sbml_model = libsbml.readSBML(sbml_path)
+        delete_ind = [
+            iSpec for iSpec in range(sbml_model.getModel().getNumSpecies())
+            if sbml_model.getModel().getSpecies(iSpec).getBoundaryCondition()
+               is True]
+        state_trajectory = np.delete(state_trajectory, delete_ind, 1)
+
+        # Convert ndarray 'state-trajectory' to data frame and save it
+        df_state_trajectory = pd.DataFrame(columns=column_names,
```
Comment (paulstapor) [inline Python_Scripts/compareStateTrajectories_JWS_1.py:181]: ?!?!?
**If I'm not mistaken, this is hogwild!**
You're taking a dataframe with state trajectories, which are reference trajectories, copy paste the columns... delete an entry... Then you simulate them with AMICI, and get the values as numeric array... prune out rows by running over the SBML species, hoping that the species ordering in the SBML file fits to the state ordering in AMICI...
... and then **glue together the state names and the numeric values from the AMICI simulation**???
... praying to heaven and hell, hoping for the orderings of these two objects to **coincide**?
... This is begging for getting errors smashed in your face!!!

File: Python_Scripts/execute_Tolerances.py [hunk 9]
```diff
@@ -20,22 +20,20 @@
...
-elif os.path.exists('../../Benchmarking_of_numerical_ODE_integration_methods/sbml2amici') and \
-        os.path.exists('../../Benchmarking_of_numerical_ODE_integration_methods/sedml_models'):
-    skip_indicator = 1
+# just for the tolerance study containing different combinations, don't change!
+study_indicator = 2
 
 # run simulation
 for iSolALg in trange(0, len(solAlg)):
     for iAtol in trange(0, len(atol)):
         for iRtol in trange(0, len(rtol)):
```
Comment (paulstapor) [inline Python_Scripts/execute_Tolerances.py:29]: This is an excellent opportunity for saying "hello" to itertools.product...

File: Python_Scripts/execute_Tolerances.py [hunk 10]
```diff
@@ -20,22 +20,20 @@
...
+            simulate(
+                atol[iAtol],
+                rtol[iRtol],
+                linSol,
+                iter,
+                solAlg[iSolALg],
+                maxStep,
+                study_indicator)
             useless_variable = os.system('clear')
             pass
```
Comment (paulstapor) [inline Python_Scripts/execute_Tolerances.py]:
```suggestion
            os.system('clear')
```

File: Python_Scripts/plot_SolAlg_Main.py [hunk 11]
```diff
@@ -30,16 +27,18 @@
...
-        list_directory_bdf.append(list_directory_general[iFile + int(len(list_directory_general)/2)])
+        list_directory_bdf.append(list_directory_general[
+            iFile + int(len(list_directory_general)/2)])
 
 # create list of doubles for scatter plot
-adams_bdf_x = []          # red
+adams_bdf_x = [] # red
 adams_bdf_y = []
-bdf_adams_x = []          # green
+bdf_adams_x = [] # green
```
Comment (paulstapor) [inline Python_Scripts/plot_SolAlg_Main.py]: Green? WTF? We don't have any green in this plot, guys!

Comment (paulstapor) [inline Python_Scripts/plot_SolAlg_Main.py]:
```suggestion
bdf_adams_x = [] # blue
```

File: Python_Scripts/plot_SolAlg_Supp1.py [hunk 12]
```diff
@@ -143,8 +166,14 @@
 # plot legend manually
 ax.plot(0.4, 25000, 'o', fillstyle='full', c='orange', markersize=marker_size)
 ax.plot(0.4, 12500, 'o', fillstyle='full', c='blue', markersize=marker_size)
-plt.text(0.6, 20000, 'AM faster: ' + str(round(len(adams_bdf_x) / len(adams_tsv_file['t_intern_ms'])*100/len(list_directory_adams), 2)) + ' %', fontsize=labelsize)
-plt.text(0.6, 10000, 'BDF faster: ' + str(round(len(bdf_adams_x)/len(adams_tsv_file['t_intern_ms'])*100/len(list_directory_adams), 2)) + ' %', fontsize=labelsize)
+plt.text(
+    0.6, 20000,
+    'AM faster: ' + str(round(len(adams_bdf_x) / len(adams_tsv_file['t_intern_ms'])*100/len(list_directory_adams), 2)) + ' %',
```
Comment (paulstapor) [inline Python_Scripts/plot_SolAlg_Supp1.py:171]: irgendwas sagt mir, dass diese Zeile jetzt nicht PEP8-konform ist... :D

File: Python_Scripts/plot_Tolerances_Main.py [hunk 13]
```diff
@@ -1,278 +1,288 @@
...
+
+###############################################################################
+# 1.PART: create Box Plot
+
+all_averaged_files = []
+tolerance_files.insert(6, '')
+tolerance_files.insert(13, '')
+tolerance_files.insert(20, '')
+tolerance_files.insert(27, '')
+tolerance_files.insert(34, '')
```
Comment (paulstapor) [inline Python_Scripts/plot_Tolerances_Main.py:54]: Is this used for something like spacing, line breaks, or similar stuff?

File: Python_Scripts/plot_NonLinSol_Supp.py [hunk 14]
```diff
@@ -20,198 +21,209 @@
...
-
-    ######## subplot 1: linear regressions of combined scatter plots
-    # list of all 14 data frames for better indexing in the future
-    all_intern_columns = [pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[]),
-                          pd.DataFrame(columns=[]), pd.DataFrame(columns=[])]
```
Comment (paulstapor) [inline Python_Scripts/plot_NonLinSol_Supp.py:33]: Warum nicht `all_intern_columns = [pd.DataFrame(columns=[])] * 14` ?

File: Python_Scripts/plot_NonLinSol_Supp.py [hunk 15]
```diff
@@ -20,198 +21,209 @@
...
+for iLinearSolverDataPoints in range(0, int(len(correct_files)/7)):
+    # concatenate Data Frames in categories of linear solver data points
+    vertically_stacked_tsv = pd.concat(
+        [all_intern_columns[7*iLinearSolverDataPoints],
+         all_intern_columns[7*iLinearSolverDataPoints + 1],
+         all_intern_columns[7*iLinearSolverDataPoints + 2],
+         all_intern_columns[7*iLinearSolverDataPoints + 3],
+         all_intern_columns[7*iLinearSolverDataPoints + 4],
+         all_intern_columns[7*iLinearSolverDataPoints + 5],
+         all_intern_columns[7*iLinearSolverDataPoints + 6]], axis=0)
```
Comment (paulstapor) [inline Python_Scripts/plot_NonLinSol_Supp.py:92]: This seems to be solvable using list comprehension....

File: Python_Scripts/plot_NonLinSol_Supp.py [hunk 16]
```diff
@@ -20,198 +21,209 @@
...
+ax1.set_xscale('log')
+ax1.set_yscale('log')
+ax1.set_xlim([0.8, 2000])
+ax1.set_ylim([0.2, 100000])
+ax1.set_xlabel('Number of state variables', fontsize=fontsize)
+ax1.set_ylabel('Simulation time [ms]', fontsize=fontsize)
+ax1.tick_params(labelsize=labelsize)
+
+linSol_for_Legend = ['Functional', 'Newton-type']
+colors = ['#e66101', '#5e3c99']
```
Comment (paulstapor) [inline Python_Scripts/plot_NonLinSol_Supp.py:82]: Why not RGBA?

File: Python_Scripts/sedml_import.py [hunk 17]
```diff
@@ -24,20 +26,32 @@ def get_sedml_infos_from_rest_api():
     return sedml_infos
 
 
-def download_all_sedml_models_from_jws(base_folder=BASE_FOLDER):
+def download_all_sedml_models_from_jws(
+        base_folder=BASE_FOLDER, only_specified=True):
```
Comment (paulstapor) [inline Python_Scripts/sedml_import.py]: Ein Docstring für `only_specified` wäre nett. Hab schon verstanden, was es tut und warum (glaube ich), aber dennoch... Ist nix, was sich *sofort* erschließt...

**Review by paulstapor — APPROVED** [review]
Overall, this looks fine to me...
Please just short answer, where comment not addressed in case.

File: .gitignore (hunk 1 above)
Comment (yannikschaelte) [reply .gitignore:3]: No, as long as there is no "/" inside the expression, the expression refers to a file in any subfolder, afaik.

File: Python_Scripts/plot_Tolerances_Main.py (hunk 13 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_Tolerances_Main.py:54]: I think so, yes.

File: Python_Scripts/compareStateTrajectories_JWS_2.py (hunk 5 above)
Comment (yannikschaelte) [reply Python_Scripts/compareStateTrajectories_JWS_2.py:166]: Si. Must be refactored. As well as repeated loading of files, not sure what's worse for performance.

File: Python_Scripts/plot_NonLinSol_Supp.py (hunk 14 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_NonLinSol_Supp.py:33]: Weil dann das passiert (python ist immer copy by reference):
```python
import pandas as pd

print("Attempt 1")

dfs = [pd.DataFrame(columns=[])] * 3
dfs[0]['a'] = [1,2]

for df in dfs:
    print(df)
    
print("Attempt 2")

dfs = [pd.DataFrame(columns=[]) for _ in range(3)]
dfs[0]['a'] = [1,2]

for df in dfs:
    print(df)
```

Gibt:
```
Attempt 1
   a
0  1
1  2
   a
0  1
1  2
   a
0  1
1  2
Attempt 2
   a
0  1
1  2
Empty DataFrame
Columns: []
Index: []
Empty DataFrame
Columns: []
Index: []

```

File: Python_Scripts/plot_NonLinSol_Supp.py (hunk 15 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_NonLinSol_Supp.py:92]: Agreed. Feel free.

File: Python_Scripts/plot_NonLinSol_Supp.py (hunk 16 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_NonLinSol_Supp.py:82]: Agreed

Comment (yannikschaelte) [reply Python_Scripts/plot_NonLinSol_Supp.py:82]: But I also like hex ;)

File: Python_Scripts/sedml_import.py (hunk 17 above)
Comment (yannikschaelte) [reply Python_Scripts/sedml_import.py]: added

File: Python_Scripts/compareStateTrajectories_JWS_1.py (hunk 7 above)
Comment (yannikschaelte) [reply Python_Scripts/compareStateTrajectories_JWS_1.py:111]: I have actually not cut out any logic I think, only
* that recursive call at the end is gone; fixed by repeatedly trying the download, as jws is sometimes unresponsive). What changed
* split up into functions
* the "maybe_download_..." function only once downloads the model to another location; thus we 1) save a factor 3 of storage space and 2) the script runs in far less time.

The json files are still saved, but  trivial to remove that in the "maybe_download_..." function.

Did I miss something?

File: Python_Scripts/compareStateTrajectories_JWS_1.py (hunk 8 above)
Comment (yannikschaelte) [reply Python_Scripts/compareStateTrajectories_JWS_1.py:181]: Yep. :see_no_evil:

File: Python_Scripts/compareStateTrajectories_JWS_2.py (hunk 6 above)
Comment (yannikschaelte) [reply Python_Scripts/compareStateTrajectories_JWS_2.py:169]: Feel free to refactor

File: Python_Scripts/plot_SolAlg_Main.py (hunk 11 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_SolAlg_Main.py]: look at it with a grain of color-blindness ;)

File: Python_Scripts/plot_SolAlg_Supp1.py (hunk 12 above)
Comment (yannikschaelte) [reply Python_Scripts/plot_SolAlg_Supp1.py:171]: Yes. Didn't want to make it too ugly on first try. But yes, should introduce variables and shorten

File: Python_Scripts/plot_SolAlg_Main.py (hunk 11 above)
Comment (paulstapor) [reply Python_Scripts/plot_SolAlg_Main.py]: I highly doubt that anyone of us suffers from Tritanopia... :D
And even then: This list is later used together with a "blue" colormap... Sorry, even if color deficient: makes no sense ;)

**Pull request merged**